flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaSinkITCase.java

```diff
@@ -0,0 +1,634 @@
+        }
+
+        @Override
+        public void copy(DataInputView source, DataOutputView target) throws IOException {}
+
+        @Override
+        public TypeSerializerSnapshot<Integer> snapshotConfiguration() {
+            return new SlowSerializerSnapshot();
+        }
+    }
+
+    /** Snapshot used in {@link FailAsyncCheckpointMapper}. */
+    public static class SlowSerializerSnapshot extends SimpleTypeSerializerSnapshot<Integer> {
+        public SlowSerializerSnapshot() {
+            super(SlowSerializer::new);
+        }
+    }
+
+    /** Fails after a checkpoint is taken and the next record was emitted. */
+    private static class FailingCheckpointMapper
+            implements MapFunction<Long, Long>, CheckpointListener, CheckpointedFunction {
+
+        private final SharedReference<AtomicBoolean> failed;
+        private final SharedReference<AtomicLong> lastCheckpointedRecord;
+
+        private volatile long lastSeenRecord;
+        private volatile long checkpointedRecord;
+        private volatile long lastCheckpointId = 0;
+        private final AtomicInteger emittedBetweenCheckpoint = new AtomicInteger(0);
+
+        FailingCheckpointMapper(
+                SharedReference<AtomicBoolean> failed,
+                SharedReference<AtomicLong> lastCheckpointedRecord) {
+            this.failed = failed;
+            this.lastCheckpointedRecord = lastCheckpointedRecord;
+        }
+
+        @Override
+        public Long map(Long value) throws Exception {
+            lastSeenRecord = value;
+            if (lastCheckpointId >= 1
+                    && emittedBetweenCheckpoint.get() > 0
+                    && !failed.get().get()) {
+                failed.get().set(true);
+                throw new RuntimeException("Planned exception.");
+            }
+            Thread.sleep(50);
+            emittedBetweenCheckpoint.incrementAndGet();
+            return value;
+        }
+
+        @Override
+        public void notifyCheckpointComplete(long checkpointId) throws Exception {
+            lastCheckpointId = checkpointId;
+            emittedBetweenCheckpoint.set(0);
+            lastCheckpointedRecord.get().set(checkpointedRecord);
+        }
+
+        @Override
+        public void snapshotState(FunctionSnapshotContext context) throws Exception {
+            checkpointedRecord = lastSeenRecord;
+        }
+
+        @Override
+        public void initializeState(FunctionInitializationContext context) throws Exception {}
+    }
+
+    private void checkProducerLeak() {
+        for (Thread t : Thread.getAllStackTraces().keySet()) {
+            if (t.getName().contains("kafka-producer-network-thread")) {
+                fail("Detected producer leak. Thread name: " + t.getName());
+            }
+        }
+    }
+
+    /**
+     * Exposes information about how man records have been emitted overall and at the beginning of a
+     * checkpoint.
+     */
+    private static final class InfiniteIntegerSource
```

Comment:
Can you just use a `env.fromSequence(0, Long.MAX_VALUE)` with a chained map that implements this functionality? You are making our future lives harder :/

Reply:
I do not see how I can easily replace it because I am relying on the fact that the Source finishes after the first `checkpointCompleted` event.

Reply:
Ah correct. Let's keep it this way then.